### PR TITLE
Spark Dataframe API

### PR DESCRIPTION
**Pull request by dvadym**
This PR implements running `DataFrame` queries for Spark `DataFrame`. It contains
1. `SparkConverter` for converting between `RDD` and `DataFrame`.
2. `Query.run_query()` implementation, which maps `Query` to `DPEngine.aggregate` call

### CODE REVIEW

File: pipeline_dp/dataframes.py [hunk 1]
```diff
@@ -34,16 +48,53 @@ class ContributionBounds:
...
+    if isinstance(df, SparkDataFrame):
+        return pipeline_dp.SparkRDDBackend(df.sparkSession.sparkContext)
+    raise NotImplementedError(
+        f"Dataframes of type {type(df)} not yet supported")
+
+
+def _create_dataframe_converter(df: SparkDataFrame) -> DataFrameConvertor:
+    """Creates a DataConvert based on type of DataFrame."""
+    if isinstance(df, SparkDataFrame):
+        return SparkConverter(df.sparkSession)
```
[the remark this answers is not in the document]
Comment (RamSaw) [reply pipeline_dp/dataframes.py:95]: ah, this is internal API, I guess then it doesn't make much difference :)

File: tests/dataframes_test.py [hunk 2]
```diff
@@ -135,15 +138,112 @@ def test_count_and_sum_query(self):
...
+            max_partitions_contributed=5,
+            max_contributions_per_partition=2,
+            min_value=min_value,
+            max_value=max_value)
+        query = dataframes.Query(df, columns, metrics_dict, bounds, public_keys)
+
+        # Act
+        query.run_query(dataframes.Budget(1, 1e-10))
+
+        # Assert
```
Comment (RamSaw) [inline tests/dataframes_test.py:195]: should we assert on the output of `run_query` as well?

File: tests/dataframes_test.py [hunk 3]
```diff
@@ -135,15 +138,112 @@ def test_count_and_sum_query(self):
...
+        # Arrange
+        spark = self._get_spark_session()
+        df = spark.createDataFrame(get_pandas_df())
+        columns = dataframes.Columns("privacy_key", "group_key", "value")
+        metrics = {
+            pipeline_dp.Metrics.COUNT: "count_column",
+            pipeline_dp.Metrics.SUM: "sum_column"
+        }
+        min_value = max_value = None
+        if pipeline_dp.Metrics.SUM in metrics:
```
Comment (RamSaw) [inline tests/dataframes_test.py:217]: this is always true

File: tests/dataframes_test.py [hunk 4]
```diff
@@ -135,15 +138,112 @@ def test_count_and_sum_query(self):
...
+        result_df = query.run_query(budget)
+
+        # Assert
+        pandas_df = result_df.toPandas()
+        pandas_df = pandas_df.sort_values(by=['group_key']).reset_index(
+            drop=True)
+        self.assertLen(pandas_df, 2)
+        # check row[0] = "key0", 0+noise, 0+noise
+        row0 = pandas_df.loc[0]
+        self.assertEqual(row0.group_key, "key0")
```
Comment (RamSaw) [inline tests/dataframes_test.py:238]: why here we access it as a field and below as a dic (`row0["count_column"]`)? if both are possible should we use the same API everywhere?

**Review by dvadym — COMMENTED** [review]
Thanks for review!

**Pull request merged**